src/entities/SubscribeCapable.ts

```diff
@@ -0,0 +1,68 @@
+import { EventEmitter, Listener, SubscriptionOptions } from './commonTypes';
+import PubNubType from 'pubnub';
+import type PubNub from '../core/pubnub-common';
+
+export abstract class SubscribeCapable {
+  protected abstract channelNames: string[];
+  protected abstract groupNames: string[];
+  protected abstract listener: Listener;
+  protected abstract eventEmitter: EventEmitter;
+  protected abstract pubnub: PubNub;
+  protected abstract options?: SubscriptionOptions;
```

Comment:
What about this and there will be no need to declare them in subclasses:
```suggestion
  protected channelNames: string[] = [];
  protected groupNames: string[] = [];
  protected listener: Listener;
  protected eventEmitter: EventEmitter;
  protected pubnub: PubNub;
  protected options?: SubscriptionOptions;
```

Reply:
The reason is, 
if we declare it like:
`protected listener: Listener;`

it forces us to make type   `Listener | undefined`

Reply:
It was a suggestion to move properties to the abstract class, so there will be no need to declare them again in all subclasses. If it is ok like we have it now, then np ;)